build.rs

```diff
@@ -218,6 +387,167 @@ fn build_msvc(target_triple: &[&str], disable_opt: bool, num_jobs: &str,
+
+    // XXX: Ideally, this would only happen for `cargo test`,
+    // but we don't know how to do that yet.
+    build_library(test_target,
+                  Vec::new().into_par_iter(),
+                  RING_TEST_SRCS,
+                  target_info,
+                  out_dir.clone(),
+                  test_header_change);
+
+    // target_vendor is only set if a nightly version of rustc is used
+    let libcxx = if target_info.target_vendor()
+        .map(|v| v == "apple")
+        .unwrap_or(target_info.target_os() == "macos" ||
+                   target_info.target_os() == "ios") ||
+                    target_info.target_os() == "freebsd" {
+        "c++"
+    } else {
+        "stdc++"
+    };
+    println!("cargo:rustc-flags=-l dylib={}", libcxx);
+    print_rerun();
+}
+
+
+fn build_library<P>(target: &Path, additional: P,
+                    lib_src: &'static [&'static str],
+                    target_info: &TargetInfo, out_dir: PathBuf,
+                    header_changed: bool)
+    where P: ParallelIterator<Item = String>
+{
+    let objs = additional.chain(lib_src.par_iter().map(|a| String::from(*a)))
+        .weight_max()
+        .map(|f| compile(&f, &target_info, out_dir.clone(), header_changed))
+        .map(|v| vec![v])
+        .reduce(Vec::new,
+                &|mut a: Vec<String>, b: Vec<String>| -> Vec<String> {
+                    a.extend(b.into_iter());
+                    a
+                });
+    if objs.par_iter()
+        .map(|f| Path::new(f))
+        .any(|p| need_run(&p, target)) {
+        let mut c = gcc::Config::new();
+
+        for f in LD_FLAGS {
+            let _ = c.flag(&f);
+        }
+        match target_info.target_os() {
+            "macos" => {
+                let _ = c.flag("-fPIC");
+                let _ = c.flag("-Wl,-dead_strip");
+            },
+            _ => {
+                let _ = c.flag("-Wl,--gc-sections".into());
+            },
+        }
+        for o in objs {
+            let _ = c.object(o);
+        }
+        c.compile(target.file_name()
+            .and_then(|f| f.to_str())
+            .expect("No filename"));
+    }
+}
+
+fn compile(file: &str, target_info: &TargetInfo, mut out_dir: PathBuf,
+           header_change: bool)
+           -> String {
+    let p = Path::new(file);
+    out_dir.push(p.file_name().expect("There is a filename"));
+    out_dir.set_extension("o");
+    if header_change || need_run(&p, out_dir.as_path()) {
+        let mut c = gcc::Config::new();
+        let _ = c.include("include");
+        match p.extension().map(|p| p.to_str()) {
+            Some(Some("c")) => {
+                for f in C_FLAGS {
+                    let _ = c.flag(f);
+                }
+            },
+            Some(Some("cc")) => {
+                for f in CXX_FLAGS {
+                    let _ = c.flag(f);
+                }
+                if target_info.target_os() != "none" {
+                    let _ = c.flag("-fstack-protector");
```

Comment:
I think you confused "CPPFLAGS" and "CXXFLAGS". This should be done for both "c" and "cc".  Here's the logic you need to implement:
```
ifneq ($(filter-out none redox,$(TARGET_SYS)),)
	CPPFLAGS += -fstack-protector
endif
```